### ISSUE TITLE
Confidence Maps Offset

### ISSUE THREAD

**Issue by roomrys** [talmolab/sleap#1611]
### Discussed in https://github.com/talmolab/sleap/discussions/1608

<div type='discussions-op-text'>

<sup>Originally posted by **daskandalis** November 25, 2023</sup>
I've often observed that the validation confidence maps are highly offset. Is this just a display issue? It's highly directional and the green (MAP?) estimate lies outside the confidence bound. 

![image](https://github.com/talmolab/sleap/assets/1887822/1a23bff5-43d0-429b-a051-fc6c703f112b)

I'm wondering though if it's related to this printout:
>INFO:sleap.nn.training:  Input shape: (1088, 1920, 1).

The width is 1088 but I think it should be 1080, the resolution of the image (unless the convolution kernels are padded in width but not height?).

Incidentally, is it possible to access the variance of the estimated position from the outputs? The measurement error would be useful in filtering.

Model parameters:
"unet": {
                "stem_stride": null,
                "max_stride": 64,
                "output_stride": 4,
                "filters": 32,
                "filters_rate": 1.5,
                "middle_block": true,
                "up_interpolate": true,
                "stacks": 1
            }

EDIT: or maybe comes from translation in the augmentation step?</div>

**Comment by talmo** [comment]
Right, I think we add padding to the image to avoid truncation when downsampling --> upsampling, but not to the confidence maps?

Relevant plotting code:

https://github.com/talmolab/sleap/blob/16241e07fb64520c8ff22d8cf180c6bf811e698a/sleap/nn/viz.py#L81-L96

Called from here during training (for single instance -- other model types plot a bit differently):

https://github.com/talmolab/sleap/blob/16241e07fb64520c8ff22d8cf180c6bf811e698a/sleap/nn/training.py#L1068-L1122

**Comment by talmo** [comment]
The viz code paths that produced this offset are gone:

- The legacy `sleap/nn/viz.py` and `sleap/nn/training.py` modules I linked above were removed entirely when the TF backend was retired (`c5a01feef`, July 2025).
- The new PyTorch-side replacement in `sleap-nn` (`sleap_nn/training/utils.py` — `plot_img` / `plot_confmaps`) draws both the image and the confmap overlay in the same coordinate space, using explicit `extent=[-0.5, W-0.5, H-0.5, -0.5]` and `output_scale = cm.H / img.H` based on the padded image that was actually fed to the model. So the cm peak and the predicted keypoint land at the same pixel — no directional offset.

If anyone hits a recurrence on current sleap-nn, please open a fresh issue at https://github.com/talmolab/sleap-nn with the training_config + a viz screenshot.

Closing as resolved.
